### ISSUE TITLE
RelationalTypeMappingSource Leak

### ISSUE THREAD

**Issue by payoff** [dotnet/efcore#32645]
We have a production application suffering from a significant memory leak. App is developed with .net 7 and published via Docker containers on Linux with PGSQL 15. 
Memory dump analysis reveals the following:

![291593676-c85e88e9-7a4b-443d-935e-11da697fa38f](https://github.com/dotnet/efcore/assets/142235/998d12d5-1542-4e65-b0bb-db2c31631048)

It should be mentioned that we use EFCoreAutoMigrator to manage 150 dynamic online tenants, each with its own DB and entities created dynamically at runtime. 
This often leads to runtime migrations and multiple configurations and load of the Models.

The leak is evident from the screenshot in the abstract class RelationalTypeMappingSource implemented by NpgsqlTypeMappingSource, there's an _explicitMappings that is never released (singletone), thus being the point where resources are not freed. Now that the problem is identified, we need to study the internal mechanisms of EFCORE to eliminate the leak.
Do you have any suggestions to share in this context?
Could a simple clear of the dictionary be a viable solution?
Thank you.
Mic

**Comment by roji** [comment]
I think the larger question here isn't really about _explicitMappings, but more about how to handle the multiple service providers (and inside them, type mapping source singletons) which are getting created in our application. In other words, if you really are creating 150 different service providers with entirely different models, it makes sense for there to be 150 type mapping sources, and in each one of them, an _explicitMappings instance.

Are you looking for an API to allow you to remove the cached service provider for a given tenant? @ajcvickers I think this was discussed recently...

**Comment by payoff** [comment]
> Are you looking for an API to allow you to remove the cached service provider for a given tenant? @ajcvickers I think this was discussed recently...

Exactly, that's precisely what we're aiming to address in our application. The issue of handling multiple tenants with their unique type mapping sources is central to our current challenges.

**Comment by roji** [comment]
Why do you think you need a unique type mapping source for each tenant?

**Comment by ajcvickers** [comment]
@payoff 

> Why do you think you need a unique type mapping source for each tenant?

Very interested to hear the answer to this, if you could share?
